app/src/main/java/com/github/raulccabreu/redukt/reducers/BaseAnnotatedReducer.kt

```diff
@@ -0,0 +1,41 @@
+package com.github.raulccabreu.redukt.reducers
+
+import com.github.raulccabreu.redukt.actions.Action
+import com.github.raulccabreu.redukt.actions.Reduce
+import java.lang.reflect.Method
+import java.security.InvalidParameterException
+import java.util.concurrent.ConcurrentHashMap
+
+
+abstract class BaseAnnotatedReducer<T> : Reducer<T> {
+
+    private val actions: ConcurrentHashMap<String, Method> = ConcurrentHashMap()
+
+    init {
+        javaClass.methods
+                .filter { it.isAnnotationPresent(Reduce::class.java) }
+                .forEach {
+                    val annotation = it.getAnnotation(Reduce::class.java) as Reduce
+
+                    if (annotation.action.isBlank())
+                        throw IllegalArgumentException("ReducerAction action cannot be empty")
+
+                    if (it.parameterTypes.size != 2)
```

Comment:
In the future we should think about a way to allow reduce actions without payload with just one argument for example. What you think @raulccabreu?

Reply:
It is good idea